readme.txt
=== FeedCache ===
Contributors: cpjolicoeur
Donate link: http://www.craigjolicoeur.com/feedcache
Tags: RSS, ATOM, feed, cache, Ruby, CRON
Requires at least: 2.3
Tested up to: 6.2.*
Stable tag: 2.0.0

Caches RSS Feeds for display on your WP sidebar.

== Requirements ==

* Ruby 3
* Rubygems with the following gems available: active_record, mysql2
* CRON access

== Description ==

FeedCache will retrieve, cache and store locally a list of RSS feeds that you can then display on your WP site. This prevents multiple HTTP requests with each page load since the feeds can be read from the cache file.

== Installation ==

1. Upload the feedcache directory to your wordpress wp-content/plugins directory. Make sure the feedcache directory is writeable by the web server (rw-rw-rw 666).

2. Change the file permissions on the master-config.txt file to rw-rw-rw- (666)

3. Activate the FeedCache plugin through your wordpress plugin menu

4. Setup FeedCache options under the Options -> FeedCache Options menu
	(a) Choose the number of feed groups you want (1 - 99)
  (b) Add your list of RSS feeds 1 per line and set the other options.  You can have up to 4 different groupings of RSS feeds to cache.
      If you wish to manually override the name of the feed, place a pipe character "|" after the RSS URL and then type the name you want to use (e.g http://www.craigjolicoeur.com|Craig P Jolicoeur)
      If you wish to manually override the number of feed items to display, place a pipe character "|" after the feed title and enter the number to display
      If you wish to manually override the feed text formatting, place a pipe character "|" after the feed display number and put either of a true/false boolean value
  (c) Take note of the "CRON Script Settings" section at the bottom - you will need this info

5. Edit the feedcache-cron.rb file
  (a) Copy the "feedcache directory" path from the CRON Script Settings into the "FEEDCACHE_DIR = ''" variable
  (b) Set the number of characters from each RSS feed you want to display

6. Add the feedcache-cron.rb ruby script to your servers CRON job (if you need help with CRON please refer
   to the following URL: http://www.unixgeeks.org/security/newbie/unix/cron-1.html)
  (e.g. 30 * * * * /usr/bin/ruby /path/to/your/wordpress/install/wp-content/plugins/feedcache/feedcache-cron.rb)

7. Add the following code to your wordpress theme where you want the RSS feed listing to be displayed

	<?php if (function_exists('feedcache_display_feeds')) { echo feedcache_display_feeds(); } ?>

	for the default (group 1) listing or you can specify a group (in this case, group 2)

	<?php if (function_exists('feedcache_display_feeds')) { echo feedcache_display_feeds(2); } ?>

	If you don't specify a group number in the function call, then Group 1 will be used.  If you want to specify the specific
	group number to display use "feedcache_display_feeds(_GROUP_NUMBER_)" where _GROUP_NUMBER_ is the number you want to display.

8. Sit back and enjoy the plugin



== Upgrading ==

**** IMPORTANT ************************************************************************
	If you are upgrading from a version of FeedCache prior to v0.9.8, please
	deactivate and delete your existing feedcache plugin directory and follow the new
	installation steps.
***************************************************************************************

1. Deactivate the previous version of FeedCache through your Wordpress plugins menu

2. Upload the new feedcache-cron.rb, feedcache.php and complete lib/ directory files to your existing feedcache directory.  You do not need to overwrite your master-config.txt file.

3. Enter the correct FEEDCACHE_DIR variable in the feedcache-cron.rb script

4. Go to Options -> FeedCache Options and update your FeedCache settings


== Frequently Asked Questions ==

= Why would I need this plugin? =

If you are using your WP installation to display other website's RSS feeds, this will save HTTP requests and improve page load times for your users.  By using a CRON job to fetch and format the feeds, the user will not have to wait during page load for the feed to be updated.

= Will FeedCache work with Atom feeds? =

Yes!  FeedCache will work with both RSS and ATOM feeds.

= Can I receive error emails from the CRON process? =

Yes, just set CRON_EMAILS = true in the feedcache-cron.rb script.  Error emails are turned off by default.

= Can I have more than 10 RSS Groups =

Yes, but you'll need to manually edit the feedcache.php file.  Find the following line:

  define("MAX_GROUPSIZE", 10);

and change the number 10 to whatever number you need.  (Just don't go overboard, :) )  Take note that when you upgrade the
FeedCache plugin in the future, you'll need to update this number every time.


== Screenshots ==

1. FeedCache Options page




CHANGELOG
== CHANGELOG ==

2.0.0 - 04/20/2023
* Upgrade to use Ruby 3
* Upgrade cron script to use all core SDK libraries (except active_record and mysql2)
* Update PHP script for current Wordpress version 6.2 and PHP 8

1.0.6.1 - 10/16/2008
* bugfix for missing false option on settings page

1.0.6 - 10/15/2008
* bugfix for truncation of feed titles during CRON job

1.0.5 - 7/13/2008
* strip non-ascii characters from title output

1.0.4 - 7/10/2008
* added plugin credits to admin page footer
* updated credential levels for adding admin subpage

1.0.3 - 07/10/2008
* update WP options to only use 2 option fields via an options array

1.0.2 - 07/09/2008
* use feed_tools gem for parsing
* single YAML config file
* store RSS cache into the WP database

0.9.8 - 05/04/2008
* allow for up to 99 feed groups now

0.9.5 - 12/02/2007
* added option to open feed links in a new browser window

0.9.1 - 12/01/2007
* bugfix for issue where files/ directory wasn't being included in the Wordpress download
* fixed issue with truncation of feed titles to include ellipses

0.9 - 11/30/2007
* allow the user to set the number of RSS groups (from 1 - 10)
* moved config and cache files to the new files/ directory
* created a new master-config file for general settings

0.8.1 - 11/24/2007
* fixed an issue with weird links using ATOM feeds from blogspot

0.8 - 11/23/2007
* now using feedparser library to parse both RSS and ATOM feeds
* cleaned up configuration so only the feedcache plugin path will need to be entered

0.7.2 - 11/20/2007
* fixed bug with config files not being set properly

0.7.1 - 11/19/2007
* rename broken screenshot file

0.7 - 11/19/2007
* Allow up to 4 different cache files to allow for multiple instances of the plugin per theme
* Fix some admin page layout issues

0.6.1 - 10/30/2007
* Add option to not send CRON error emails

0.6 - 10/01/2007
* Fix broken feed formatting options

0.5.5 - 09/27/2007
* Added user-agent to the CRON script to force RSS-only feeds from feedburners SmartFeed service

0.5.2 - 09/24/2007
* Lower the error output level from the CRON job again

0.5.1 - 09/24/2007
* Lower the error output level from the CRON job

0.5 - 09/15/2007
* Use a temp file when grabbing new cache

0.4 - 09/13/2007
* Updates to the Ruby CRON script for better error handling

0.3.1 - 09/12/2007
* Fixed bug that would cause PHP header errors during login/logout

0.3 - 09/04/2007
* Added option to manually override the number of feed items to display on a per feed basis
* Added option to manually override the feed text formatting option on a per feed basis

0.2 - 09/03/2007
* Updated README file
* Added option for manually named feed titles

0.1 - 08/31/2007
* Initial Release


feedcache-cron.rb
#!/usr/bin/env ruby

# Add the path to your feedcache directory here
# FEEDCACHE_DIR = '/src/cpjolicoeur/feedcache/feedcache'
FEEDCACHE_DIR = '/Users/cpjolicoeur/src/feedcache/feedcache'
# How many characters from each feed item do you want to display
CHAR_COUNT = 60
# Set to 'true' if you want to receive error emails from the CRON job
CRON_EMAILS = true

#################################################################
#                                                               #
#  DO NOT EDIT BELOW THIS LINE                                  #
#                                                               #
#################################################################
$LOAD_PATH << File.expand_path(__dir__)

require 'active_record'
require 'rss'
require 'open-uri'
require 'yaml'

# Read master config settings
MASTER_CONFIG = "#{FEEDCACHE_DIR}/master-config.txt"
params = []
cfg = File.open(MASTER_CONFIG, 'r') do |f|
  params = f.gets.split('~');
end

# parse the parameters from the config file
@groups_num   = params[0].strip.to_i
@display_num  = params[1].strip
@title_pre    = params[2].strip
@title_post   = params[3].strip
@format_text  = params[4].strip == 'true' ? true : false
@link_target  = params[5].strip == 'true' ? '_blank' : '_self'
WPDB_PREFIX   = params[6].strip
@wpdb_host    = params[7].strip
@wpdb_name    = params[8].strip
@wpdb_user    = params[9].strip
@wpdb_pass    = params[10].strip

# Load config and cache file variables
CONFIG_FILE = "#{FEEDCACHE_DIR}/feedcache-config.yml"

ActiveRecord::Base.establish_connection(
  :adapter  => 'mysql2',
  :socket   => '/var/run/mysqld/mysqld.sock',
  :host     => @wpdb_host,
  :username => @wpdb_user,
  :password => @wpdb_pass,
  :database => "#{@wpdb_name}\0"
)

class WPFeed < ActiveRecord::Base
  self.table_name = "#{WPDB_PREFIX}feedcache_data"
  self.primary_key = 'id'
end

class String
  def to_ascii_iconv
    self.encode('ASCII', invalid: :replace, undef: :replace, replace: '')
  end
end

# RSS formatting function
def shorten_text(txt)
  if txt.size > CHAR_COUNT
    text = "#{txt} ".slice(0,CHAR_COUNT)
    # need to break on the last space
    if text.include?(' ') and text.slice(text.size-1, 1) != ' '
      text = text.slice(0, text.size - (text.reverse.index(' ') + 1))
      text << '...'
    end
    return text
  else
    return txt
  end
end

def fetch_rss(url)
  content = URI.open(url).read
  rss = RSS::Parser.parse(content, validate: false)
  rss
rescue StandardError => e
  puts "#{e.class} => Error fetching or parsing #{url}: #{e.message}"
  nil
end

begin # read the config file settings
  @all_feeds = {}
  yaml_config = YAML.load_file(CONFIG_FILE)
  1.upto(@groups_num) do |num|
    feeds = []
    next if yaml_config["group#{num}"].nil?
    yaml_config["group#{num}"].split(/\n/).each {|x| feeds << x.strip if (!x.nil? && !x.strip.blank?) }
    @all_feeds[num] = feeds
    feeds = nil
  end
rescue => e
  if CRON_EMAILS
    puts "Error reading YAML configuration file: #{e.message}"
    puts e.inspect
    puts e.backtrace
  end
end

# puts "\n\n**ALL FEEDS**\n\n#{@all_feeds}\n\n"

@all_feeds.each do |k,v|
  tmp = ''
  @processed = 0

  v.each do |feed|
    html_text = ''
    data = feed.split('|')
    feed_url, feed_title, feed_num, feed_format = data[0], data[1], data[2], data[3]
    begin
      rss = fetch_rss(feed_url)
      next unless rss

      feed_items = []
      if 'rss' == rss.feed_type
        feed_items = rss.channel.items
        feed_title = feed_title || rss.channel.title || ''
      elsif 'atom' == rss.feed_type
        feed_items = rss.entries
        feed_title = feed_title || rss.title || ''
      end

      html_text << @title_pre + feed_title + @title_post
      html_text << "<ul>"

      feed_items.each_with_index do |item, idx|
        break if feed_num ? feed_num.to_i == idx.to_i : @display_num.to_i == idx.to_i
        output = ''
        output << "<li><a href='#{item.link}' target='#{@link_target}'>"
        item_title = item.title.to_s.to_ascii_iconv
        if feed_format && feed_format == 'true'
          txt = "#{item_title.downcase.gsub(/^[a-z]|\s+[a-z]/) {|a| a.upcase}}"
          output << shorten_text(txt)
        elsif @format_text == true
          txt = "#{item_title.downcase.gsub(/^[a-z]|\s+[a-z]/) {|a| a.upcase}}"
          output << shorten_text(txt)
        else
          output << "#{item_title}"
        end
        output << "</a></li>\n"
        html_text << output
      end # end feed_items.each_with_index

      html_text << "</ul><br />\n"
      tmp << html_text
      @processed += 1
    rescue => e
      if CRON_EMAILS
        puts "Error processing feed - Group #{k.to_i + 1} - #{feed_url}"
        puts e.inspect
        puts e.backtrace
      end
    end # end begin...rescue
  end # end v.each do |feed|

  # if we had new feeds, move them to the cache file
  if @processed > 0
    puts "****** Group id: #{k}\nData: #{tmp}\n"
    # wp_data = WPFeed.find_or_initialize_by(group_id: k)
    # wp_data.update(data: tmp, updated_at: Time.now)
  end
end # end @all_feeds.each do |k,v|



feedcache-lite.rb
# TODO: update this to use the new config file setup and memory management code

$LOAD_PATH << File.expand_path(File.dirname(__FILE__))

require 'optparse'
require 'net/http'
require 'lib/feedparser'
require 'uri'
require 'yaml'

# Default options
options = {
  :display_num => 5,
  :title_pre => "<h3>",
  :title_post => "</h3>",
  :format_text => 'true',
  :link_target => '_blank',
  :char_count => 75, 
  :group => 1,
  :cron_emails => false
}
OptionParser.new do |opts|
  opts.banner = "Usage: feedcache-lite.rb [options]"

  opts.on("-v", "--[no-]verbose", "Run verbosely") do |v|
    options[:verbose] = v
  end
  
  opts.on("-e", "--cron-emails", "Send CRON emails") do |e|
    options[:cron_emails] = e
  end
  
  opts.on("-p PATH", "--path PATH", "Path of config file") do |p|
    options[:path] = p
  end
  
  opts.on("-g GROUP", "--group GROUP", "Group number to process") do |g|
    options[:group] = g.to_i
  end
  
  opts.on("-n NUM", "--num NUM", "Number of entries to show") do |n|
    options[:display_num] = n.to_i
  end
  
  opts.on("-r PRE", "--pre PRE", "Title pre text") do |r|
    options[:title_pre] = r
  end
  
  opts.on("-s POST", "--post POST", "Title post text") do |s|
    options[:title_post] = s
  end
  
  opts.on("-f FORMAT", "--format FORMAT", ['true', 'false'], "Format text [true|false]") do |f|
    options[:format_text] = f
  end
  
  opts.on("-l LINK", "--link LINK", "HTML link target") do |l|
    options[:link_target] = l
  end
  
  opts.on("-c COUNT", "--count COUNT", "Character count") do |c|
    options[:char_count] = c.to_i
  end
end.parse!

raise "You must specify at least the -g group number option" if options[:group].nil?

CONFIG_FILE = "#{options[:path]}"
CACHE_FILE = CONFIG_FILE.gsub(/config/, 'cache')

# RSS formatting function
def shorten_text(txt, char_count)
  if txt.size > char_count
    text = "#{txt} ".slice(0,char_count)
    # need to break on the last space
    if text.include?(' ') and text.slice(text.size-1, 1) != ' '
      text.slice!(0, text.size - (text.reverse.index(' ') + 1))
      text << '...'
    end
    return @text
  else
    return txt
  end
end

begin # read the config file settings
  feeds = []
  yaml_config = YAML.load_file(CONFIG_FILE)
  yaml_config["group#{options[:group]}"].each { |x| feeds << x.strip if (!x.nil? && !x.strip.blank?) }
  feeds = nil
rescue => e
  if options[:cron_emails]
    puts "Error reading configuration file"
    puts e.inspect
    puts e.backtrace
  end
end  

tmp = ''
@processed = 0

# parse the feeds here
feeds.each do |feed|
  html_text = ''
  data = feed.split('|')
  feed_url, feed_title, feed_num, feed_format = data[0], data[1], data[2], data[3]
  begin
    source = Net::HTTP::get URI::parse(feed_url)
    fp = FeedParser::Feed::new(source)
      html_text << options[:title_pre] + (feed_title || fp.title) + options[:title_post]
      html_text << "<ul>"
      fp.items.each_with_index do |item, idx|
        break if feed_num ? feed_num.to_i == idx.to_i : options[:display_num].to_i == idx.to_i
        output = ''
        output << "<li><a href='#{item.link}' target='#{options[:link_target]}'>"
        if feed_format && feed_format == 'true'
          txt = "#{item.title.downcase.gsub(/^[a-z]|\s+[a-z]/) {|a| a.upcase}}"
          output << shorten_text(txt, options[:char_count])
        elsif options[:format_text] == 'true'
          txt = "#{item.title.downcase.gsub(/^[a-z]|\s+[a-z]/) {|a| a.upcase}}"
          output << shorten_text(txt, options[:char_count])
        else
          output << "#{item.title}"
        end
        output << "</a></li>\n"
        html_text << output
      end # end fp.items.each
      html_text << "</ul><br />\n"
      tmp << html_text
      @processed += 1
  rescue => e
    puts "Error processing feed - #{feed_url}"
    puts e.inspect
    puts e.backtrace
  end  
end

# if we had new feeds, move them to the cache file
if @processed > 0
  cache = File::open(CACHE_FILE, "w")
  cache << tmp
  cache.close
end


lib/feedparser/textconverters.rb
# for URI::regexp
require 'uri'
require 'lib/feedparser/html2text-parser'

# This class provides various converters
class String
  # is this text HTML ? search for tags. used by String#text2html
  def html?
    return (self =~ /<p>/i) || (self =~ /<\/p>/i) || (self =~ /<br>/i) || (self =~ /<br\s*(\/)?\s*>/i) || (self =~ /<\/a>/i) || (self =~ /<img.*>/i)
  end

  # returns true if the text contains escaped HTML (with HTML entities). used by String#text2html
  def escaped_html?
    return (self =~ /&lt;img src=/i) || (self =~ /&lt;a href=/i) || (self =~ /&lt;br(\/| \/|)&gt;/i) || (self =~ /&lt;p&gt;/i)
  end

  def escape_html
    r = self.gsub('&', '&amp;')
    r = r.gsub('<', '&lt;')
    r = r.gsub('>', '&gt;')
    r
  end

MY_ENTITIES = {}
FeedParser::HTML2TextParser::entities.each do |k, v|
  MY_ENTITIES["&#{k};"] = [v].pack('U*')
  MY_ENTITIES["&##{v};"] = [v].pack('U*')
end

  # un-escape HTML in the text. used by String#text2html
  def unescape_html
    r = self
    MY_ENTITIES.each do |k, v|
      r = r.gsub(k, v)
    end
    r
  end

  # convert text to HTML
  def text2html(feed)
    text = self.clone
    realhtml = text.html?
    eschtml = text.escaped_html?
    # fix for RSS feeds with both real and escaped html (crazy!):
    # we take the first one
    if (realhtml && eschtml)
      if (realhtml < eschtml)
        eschtml = nil
      else
        realhtml = nil
      end
    end
    if realhtml
      # do nothing
    elsif eschtml
      text = text.unescape_html
    else
      # paragraphs
      text.gsub!(/\A\s*(.*)\Z/m, '<p>\1</p>')
      text.gsub!(/\s*\n(\s*\n)+\s*/, "</p>\n<p>")
      # uris
      text.gsub!(/(#{URI::regexp(['http','ftp','https'])})/,
          '<a href="\1">\1</a>')
    end
    # Handle broken hrefs in <a> and <img>
    if feed and feed.link
      text.gsub!(/(\s(src|href)=['"])([^'"]*)(['"])/) do |m|
        begin
          first, url, last = $1, $3, $4
          if (url =~ /^\s*\w+:\/\//) or (url =~ /^\s*\w+:\w/)
            m
          elsif url =~ /^\//
            (first + feed.link.split(/\//)[0..2].join('/') + url + last)
          else
            t = feed.link.split(/\//)
            if t.length == 3 # http://toto with no trailing /
              (first + feed.link + '/' + url + last)
            else
              if feed.link =~ /\/$/
                (first + feed.link + url + last)
              else
                (first + t[0...-1].join('/') + '/' + url + last)
              end
            end
          end
        rescue
          m
        end
      end
    end
    text
  end

  # Remove white space around the text
  def rmWhiteSpace!
    return self.gsub!(/\A\s*/m, '').gsub!(/\s*\Z/m,'')
  end

  # Convert a text in inputenc to a text in UTF8
  # must take care of wrong input locales
  def toUTF8(inputenc)
    if inputenc.downcase != 'utf-8'
      # it is said it is not UTF-8. Ensure it is REALLY not UTF-8
      begin
        if self.unpack('U*').pack('U*') == self
          return self
        end
      rescue
        # do nothing
      end
      begin
        return self.unpack('C*').pack('U*')
      rescue
        return self #failsafe solution. but a dirty one :-)
      end
    else
      return self
    end
  end
end


feedcache.php
<?php
/*
Plugin Name: FeedCache
Plugin URI: http://www.craigjolicoeur.com/feedcache
Description: Caches RSS Feeds for display on WP site sidebar.  This prevents multiple HTTP requests with each page load since the feeds can be read from the cache file.
Author: Craig P Jolicoeur
Version: 2.0.0
Author URI: http://www.craigjolicoeur.com/
*/

//***************************************************
//
//         DO NOT EDIT BELOW THIS LINE
//
//***************************************************

// Pre-2.6 compatibility
if ( !defined('WP_CONTENT_URL') ) {
	define('WP_CONTENT_URL', get_option('siteurl') . '/wp-content');
}
if ( !defined('WP_CONTENT_DIR') ) {
	define('WP_CONTENT_DIR', ABSPATH . 'wp-content');
}

// Constants
define("FEEDCACHE_PATH", WP_CONTENT_DIR.'/plugins/'.plugin_basename(dirname(__FILE__)));
define("FEEDCACHE_FILES_PATH", FEEDCACHE_PATH . '/');
define("DEFAULT_PRE_TAG", "<h3>");
define("DEFAULT_POST_TAG", "</h3>");
define("DEFAULT_FORMAT_TEXT", "true");
define("DEFAULT_GROUP_NUM", '4');
define("DEFAULT_DISPLAY_NUM", '5');
define("CHAR_COUNT", 75);
define("MAX_GROUPSIZE", 10);

// Include Spyc PHP YAML library
include_once(FEEDCACHE_PATH . '/lib/spyc/spyc.php');

// Wordpress hooks
add_action('admin_menu', 'feedcache');
register_activation_hook(__FILE__, 'feedcache_install');

// DB setup
$feedcache_db_version = "1.0";
function feedcache_install () {
	global $wpdb;
	global $feedcache_db_version;

	// Install DB table if not already installed
	$table_name = $wpdb->prefix . "feedcache_data";
	if ($wpdb->get_var("show tables like '$table_name'") != $table_name) {
		$sql = "CREATE TABLE " . $table_name . " (
			id mediumint(9) NOT NULL AUTO_INCREMENT,
			group_id mediumint(9) NOT NULL,
			data text default NULL,
			updated_at datetime default NULL,
			UNIQUE KEY id (id),
			UNIQUE KEY group_id (group_id)
			);";

			require_once(ABSPATH . 'wp-admin/includes/upgrade.php');
			dbDelta($sql);

			add_option("feedcache_db_version", $feedcache_db_version);
	}

	// Check if DB table upgrade is needed
	$installed_ver = get_option("feedcache_db_version");
	if ($installed_ver != $feedcache_db_version) {
		$sql = ""; // New DB schema here

		require_once(ABSPATH . 'wp-admin/inclues/upgrade.php');
		dbDelta($sql);

		update_option("feedcache_db_version", $feedcache_db_verion);
	}

	$new_groups = array();
	$new_options = array(
		'group_num' => DEFAULT_GROUP_NUM,
		'display_num' => DEFAULT_DISPLAY_NUM,
		'title_pre' => DEFAULT_PRE_TAG,
		'title_post' => DEFAULT_POST_TAG,
		'format_text' => DEFAULT_FORMAT_TEXT,
		'target_blank' => false
	);
	// if old options exist, update to new system
	foreach ( $new_options as $key => $value ) {
		if( $existing = get_option('feedcache_'.$key) ) {
			$new_options[$key] = $existing;
			delete_option('feedcache_'.$key);
		}
	}
	// if old groups exist, update to new system
	for ($i=1; $i<=99; $i++) {
		if( $existing = get_option("feedcache_rss_list$i") ) {
			$new_groups["group$i"] = $existing;
			delete_option("feedcache_rss_list$i");
		}
	}
	add_option('plugin_feedcache_groups', $new_groups);
	add_option('plugin_feedcache_options', $new_options);

}

// Functions
function feedcache_display_feeds($rss_group = '1', $fname = 'feedcache-cache') {
	global $wpdb;
	$feed_data = $wpdb->get_row("SELECT * FROM " . $wpdb->prefix . "feedcache_data WHERE group_id = $rss_group");
	return $feed_data->data;
}

function fc_build_config_file($group_array, $fname = 'feedcache-config') {
		$fpath = FEEDCACHE_FILES_PATH . "$fname.yml";
		$yaml = Spyc::YAMLDump($group_array);

		// create the config file if it doesn't exist and make it writeable
		if (!file_exists($fpath)) {
			$tmp_handle = fopen($fpath, 'w') or die("can't create file");
			fclose($tmp_handle);
			chmod($fpath, 0666);
		}

	  if (is_writable($fpath)) {
	    if (!$handle = fopen($fpath, "w")) {
	      echo "Cannot open file ($fpath)";
	      exit;
	    }
	    if (fwrite($handle, "$yaml") === FALSE) {
	      echo "Cannot write to file ($fpath)";
	      exit;
	    }
	    fclose($handle);
	  } else {
	    echo "The config file ($fpath) is not writable";
	  }
}

function fc_build_master_config($group_num, $display_num, $title_pre, $title_post, $format_text, $target_blank, $fname = 'master-config') {
	global $wpdb;
	$fpath = FEEDCACHE_PATH . "/$fname" . ".txt";

  if (is_writable($fpath)) {
    if (!$handle = fopen($fpath, "w")) {
      echo "Cannot open file ($fpath)";
      exit;
    }
    if (fwrite($handle, "$group_num~$display_num~$title_pre~$title_post~$format_text~$target_blank~$wpdb->prefix~".DB_HOST."~".DB_NAME."~".DB_USER."~".DB_PASSWORD) === FALSE) {
      echo "Cannot write to file ($fpath)";
      exit;
    }
    fclose($handle);
  } else {
    echo "The master-config file ($fpath) is not writable";
  }
}

function feedcache() {
	if ( current_user_can('switch_themes') ) {
    add_submenu_page('plugins.php', 'FeedCache Options', 'FeedCache', 'switch_themes', __FILE__, 'feedcache_subpanel');
	}
}

function feedcache_admin_footer() {
	$plugin_data = get_plugin_data( __FILE__ );
	printf('%1$s plugin | Version %2$s | by %3$s<br />', $plugin_data['Title'], $plugin_data['Version'], $plugin_data['Author']);
}

function feedcache_subpanel() {

	add_action('in_admin_footer', 'feedcache_admin_footer');

	if ($_POST['stage'] == 'prep') {
		$options = get_option('plugin_feedcache_options');
		$options['group_num'] = $_POST['feedcache_group_num'];
		update_option('plugin_feedcache_options', $options);
	}

	$options = get_option('plugin_feedcache_options');
	$number_of_groups = $options['group_num'];

  if ($_POST['stage'] == 'process' ) {
		for ($i=1; $i<=$number_of_groups; $i++) {
			$group_array["group$i"] = $_POST["feedcache_rss_list$i"]; //explode("\n", $_POST["feedcache_rss_list$i"]);
		}
		update_option('plugin_feedcache_groups', $group_array);
		fc_build_config_file($group_array);

		// update other variables and master config
		$new_options = $_POST['feedcache'];
		$new_options['group_num'] = $number_of_groups;
		update_option('plugin_feedcache_options', $new_options);
		fc_build_master_config($new_options['group_num'], $new_options['display_num'], $new_options['title_pre'], $new_options['title_post'], $new_options['format_text'], $new_options['target_blank']);
  }

	$options = get_option('plugin_feedcache_options');
	$groups = get_option('plugin_feedcache_groups');
?>

   <div class="wrap">
       <h2 id="write-post">FeedCache&hellip;</h2>
       <p>Fill in the list of RSS feeds you want to process and how you want them formatted and displayed on your site.  If you notice any bugs or have suggestions please contact the developers at <a href="http://www.craigjolicoeur.com/feedcache" title="Craig Jolicoeurs's Home Page">FeedCache on craigjolicoeur.com</a>.</a></p>
       <form method="post" action="<?php echo $_SERVER['REQUEST_URI']; ?>" name="prep_form">
				<table class="form-table">
					<tbody>
						<tr valign="top">
							<th scope="row">How many RSS Groups do you want?</th>
							<td>
								<label for="feedcache_group_num">
									<select name="feedcache_group_num" onchange="document.prep_form.submit();">
										<?php
											for($i=1; $i <= MAX_GROUPSIZE; $i++) {
												if($i == $number_of_groups) {
													echo "<option value='$i' selected>$i</option>";
												}
												else {
													echo "<option value='$i'>$i</option>";
												}
											}
										?>
									</select>
								</label>
							</td>
						</tr>
					</tbody>
				</table>
         <input type="hidden" name="stage" value="prep" />
			</form><!-- end #prep_form -->

       <form method="post" action="<?php echo $_SERVER['REQUEST_URI']; ?>">
		  	<div>
					<p>Enter the list of RSS Feeds [1 per line]</p>
				  <p>Following the RSS URL, there are up to three (3) additional options you can specify for each feed: 1. Feed title, 2.
				  Number of items to display, 3. Option to enable special formatting of feed text (boolean value - true/false)</p>
				  <p>If you wish to specify the third option, you MUST also set the first two options.  If you wish to set the
				  second option, you MUST also specify the first option</p>
				  <p>All options should be separated by the pipe ('|') character with no spaces before and after the pipe</p>
		      <p>(e.g. http://www.yourfeed.com/feed|Feed Title|4|false)</p>
				</div>
				<table class="form-table">
					<tbody>
						<?php
							for ($i=1; $i<=$number_of_groups; $i++) {
						?>
							<tr valign="top">
								<th scope="row">Group <?php echo $i; ?></th>
								<td>
									<textarea name="feedcache_rss_list<?php echo $i; ?>" rows="10" cols="60" style="width:98%;"><?php echo $groups["group$i"]; ?></textarea>
								</td>
							</tr>
						<?php
							}
						?>
						<tr valign="top">
							<th scope="row">Number of articles to display:</th>
							<td><input type="text" name="feedcache[display_num]" style="width:65px;" value="<?php echo $options['display_num']; ?>" />
							</td>
						</tr>
						<tr valign="top">
							<th scope="row">Format feed text for capitalization: </th>
							<td>
								<label for"feedcache[format_text]">
									<select name="feedcache[format_text]">
								    <option value='true' <?php if ($options['format_text'] == 'true') { echo 'selected'; } ?> >True</option>
                    <option value='false' <?php if ($options['format_text'] == 'false') { echo 'selected'; } ?> >False</option>
                  </select>
								</label>
							</td>
						</tr>
						<tr valign="top">
							<th scope="row">Open feed links in a new window:</th>
							<td>
								<label for"feedcache[target_blank]">
									<select name="feedcache[target_blank]">
								    <option value='true' <?php if ($options['target_blank'] == 'true') { echo 'selected'; } ?> >True</option>
                    <option value='false' <?php if ($options['target_blank'] == 'false') { echo 'selected'; } ?> >False</option>
                  </select>
								</label>
							</td>
						</tr>
						<tr valign="top">
							<th scope="row">Feed title display pre/post tags:</th>
							<td>
                 <input type="text" name="feedcache[title_pre]" style="width:65px;" value=<?php echo '\''. $options['title_pre'] . '\''; ?> />&nbsp;/&nbsp;
                 <input type="text" name="feedcache[title_post]" style="width:65px;" value=<?php echo '\''. $options['title_post'] . '\''; ?> /> <em style="font-size:11px;">e.g &lt;h2&gt; / &lt;/h2&gt;</em>
							</td>
						</tr>
					</tbody>
				</table>

	      <div class="submit">
					<input type="submit" value="Update FeedCache Preferences &raquo;" name="Submit" />
				</div>
				<input type="hidden" name="stage" value="process" />
			</form><!-- end main form -->


			<div class="options" id="script-settings">
        <h3>CRON Script Settings</h3>
				<p>
			    Here is the <b>feedcache directory path</b> to use in the CRON script:<br />
			    <?php echo FEEDCACHE_PATH . "<br />"; ?>
				</p>
			</div>
   </div><!-- end .wrap -->

<?php
}
?>
